flow/layers/opacity_layer.cc

```diff
@@ -4,65 +4,120 @@
 
-void OpacityLayer::EnsureSingleChild() {
-  FML_DCHECK(layers().size() > 0);  // OpacityLayer should never be a leaf
+#if !defined(OS_FUCHSIA)
+void OpacityLayerBase::Preroll(PrerollContext* context,
+                               const SkMatrix& matrix) {
+  const SkAlpha alpha = SkColorGetA(color());
+  const float parent_is_opaque = context->is_opaque;
 
-  if (layers().size() == 1) {
-    return;
-  }
+  context->mutators_stack.PushOpacity(alpha);
+  context->is_opaque = parent_is_opaque && (alpha == 255);
+  ContainerLayer::Preroll(context, matrix);
+  context->is_opaque = parent_is_opaque;
+  context->mutators_stack.Pop();
+}
+#endif
 
-  // Be careful: SkMatrix's default constructor doesn't initialize the matrix to
-  // identity. Hence we have to explicitly call SkMatrix::setIdentity.
-  SkMatrix identity;
-  identity.setIdentity();
-  auto new_child = std::make_shared<flutter::TransformLayer>(identity);
+OpacityLayer::OpacityLayer(SkAlpha opacity, const SkPoint& offset)
+    : OpacityLayerBase(SK_ColorBLACK,
+                       opacity,
+                       kOpacityElevationWhenUsingSystemCompositor),
+      offset_(offset) {
+  // Ensure OpacityLayer has only one direct child.
+  //
+  // This is needed to ensure that retained rendering can always be applied to
+  // save the costly saveLayer.
+  //
+  // Any children will be actually added as children of this empty
+  // ContainerLayer.
+  ContainerLayer::Add(std::make_shared<ContainerLayer>());
+}
 
-  for (auto& child : layers()) {
-    new_child->Add(child);
-  }
-  ClearChildren();
-  Add(new_child);
+void OpacityLayer::Add(std::shared_ptr<Layer> layer) {
+  GetChildContainer()->Add(std::move(layer));
 }
 
 void OpacityLayer::Preroll(PrerollContext* context, const SkMatrix& matrix) {
-  EnsureSingleChild();
+  ContainerLayer* container = GetChildContainer();
+  FML_DCHECK(!container->layers().empty());  // OpacityLayer can't be a leaf.
+
+  // Factor in the offset during Preroll.  |OpacityLayerBase| will handle the
+  // opacity.
   SkMatrix child_matrix = matrix;
   child_matrix.postTranslate(offset_.fX, offset_.fY);
   context->mutators_stack.PushTransform(
       SkMatrix::MakeTrans(offset_.fX, offset_.fY));
-  context->mutators_stack.PushOpacity(alpha_);
-  ContainerLayer::Preroll(context, child_matrix);
-  context->mutators_stack.Pop();
+  OpacityLayerBase::Preroll(context, child_matrix);
   context->mutators_stack.Pop();
-  set_paint_bounds(paint_bounds().makeOffset(offset_.fX, offset_.fY));
-  // See |EnsureSingleChild|.
-  FML_DCHECK(layers().size() == 1);
-  if (!context->has_platform_view && context->raster_cache &&
-      SkRect::Intersects(context->cull_rect, paint_bounds())) {
-    Layer* child = layers()[0].get();
-    SkMatrix ctm = child_matrix;
+
+  // When using the system compositor, do not include the offset or use the
+  // raster cache, since we are rendering as a separate piece of geometry.
```

Comment:
Why should be not cache the rendered content in the raster cache? Presumably the for the same parameters, system compositor should be rendering the same content.

Reply:
Ahh, this makes sense.  I'll restore the raster cache code.

The code for Vulkan surface re-use i moved to FuchsiaSystemCompositedContainerLayer, so any layer (including opacity) that system composites should benefit from that now.

So, we are caching at both levels - the raster cache will cache the results of Paint(), and the retained node will cache the results of UpdateScene().